### PR TITLE
Assignment3base

### PR DESCRIPTION
**Pull request by 0xlbr**

### CODE REVIEW

File: assignment3/TestWordSearch.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		wordList.add("card");
+		wordList.add("cat");
+		wordList.add("cart");
+		d = new Dictionary(wordList);
+		char [] firstRow= {'a', 'a', 'r'};
+		char [] secondRow= {'t', 'c', 'd'};
+		grid = new char [2] [];
+		grid [0] = firstRow;
+		grid [1] = secondRow;
+	}
```
Comment (simon-frankau) [inline assignment3/TestWordSearch.java:29]: There are nicer initializer formats that you can use. I suggest having a quick read up on the options.

File: assignment3/TestWordSearch.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	
+	@Test
+	public void testExample() {
+		HashSet<String> validWords = WordSearch.wordsInGrid(2, 3, grid, d);
+		LinkedHashSet<String> orderedWords = new LinkedHashSet<String>(validWords);
```
Comment (simon-frankau) [inline assignment3/TestWordSearch.java:34]: While LinkedHashSet knows the order items are inserted into it, it will insert them into the order it gets them from validWords. In other words, I don't believe that using LinkedHashSet actually improves ordering for you.

File: assignment3/TestWordSearch.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		LinkedHashSet<String> expectedWords = new LinkedHashSet<String>();
+		expectedWords.add("cat");
+		expectedWords.add("car");
+		expectedWords.add("card");
+		assertEquals(orderedWords, expectedWords);
```
Comment (simon-frankau) [inline assignment3/TestWordSearch.java:39]: There are some unit test helpers that should allow you to do set (order-invariant) comparison more easily.

File: assignment3/TestWordSearch.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	}
+}
```
Comment (simon-frankau) [inline assignment3/TestWordSearch.java:41]: This is somewhat limited testing. Try to test a few corner cases. Don't forget to test repeated characters - you want to make sure that no character in the grid is used more than once.

File: assignment3/WordSearch.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+import javax.print.attribute.Size2DSyntax;
+
+
+public class WordSearch {
+
+/**
+* Call a method that starts a search for valid words starting from each cell on the grid
+*/
+public static HashSet<String> wordsInGrid(int rows, int cols, char [] [] grid, Dictionary dictionary){
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:17]: "rows" and "cols" can be inferred from "grid".

File: assignment3/WordSearch.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	
+	HashSet<String> validWords = new HashSet<String>();
+	for (int i = 0; i < rows; i++) {
+		for (int j = 0; j < cols; j++) {
+			/** the array keeps track of visited cells when starting from each cell */
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:22]: "/**" is for a javadoc comment - is that your intention here?

File: assignment3/WordSearch.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+			boolean [] [] visitedFromCell = new boolean [rows] [cols];
+			/** stores valid words starting from each cell */
+			HashSet<String> wordsFromCell = searchFromCell(i, j,"", grid,visitedFromCell, dictionary, rows, cols);
+				if (!wordsFromCell.isEmpty()){
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:26]: Why check? If you add an empty set to another set, that's a perfectly safe no-op.

File: assignment3/WordSearch.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	/** add the current letter to the prefix */
+	String temp = word + grid [i] [j];
+	visited [i] [j] = true;
+	
+	/**if temp is a valid word add it to the set of words*/
+	if (dictionary.isWord(temp)) {
+		wordsFromCell.add(temp);
+	
+	/** if temp is a valid prefix iterate in all 8 possible directions starting from current cell */	
+	}if (dictionary.isPrefix(temp)) {
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:51]: Put a newline after the first "}".

File: assignment3/WordSearch.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		for (int k = i-1; k < i+2 && k<rows; k++) {
+			for (int q = j-1; q < j+2 && q<cols; q++) {
+				/** if the cell fits in the grid and hasn't been visited yet call searchFromCell recursively*/
+				if (q>-1 && k>-1 && !visited [k] [q] ) {
+					wordsFromCell.addAll(searchFromCell(k, q, temp, grid, visited, dictionary, rows, cols));
+					/** restore the cell as unvisited */
+					visited [k] [q] = false;
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:58]: I don't like setting visited and resetting is in different calls. It's much easier to maintain the invariant if you have the same call setting and resetting it.

i.e. just write "visited[i][k]" just before the "return".

If you did this, you'd not need to create a new "visitedFromCell" in each iteration of the loop in "wordsInGrid", since each call to "searchFromCell" would leave visited exactly how it found it.

File: assignment3/WordSearch.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	/**if temp is a valid word add it to the set of words*/
+	if (dictionary.isWord(temp)) {
+		wordsFromCell.add(temp);
+	
+	/** if temp is a valid prefix iterate in all 8 possible directions starting from current cell */	
+	}if (dictionary.isPrefix(temp)) {
+		for (int k = i-1; k < i+2 && k<rows; k++) {
+			for (int q = j-1; q < j+2 && q<cols; q++) {
+				/** if the cell fits in the grid and hasn't been visited yet call searchFromCell recursively*/
+				if (q>-1 && k>-1 && !visited [k] [q] ) {
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:55]: Your use of spaces around operators doesn't match up with the Google Java code style. Please use the style - it's easier to read, and will make your code consistent (e.g. "q < j+2" vs. "q>-1").

File: assignment3/WordSearch.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+					wordsFromCell.addAll(searchFromCell(k, q, temp, grid, visited, dictionary, rows, cols));
+					/** restore the cell as unvisited */
+					visited [k] [q] = false;
+				}	
+			}
+		}
+	}
+	return wordsFromCell;
+}
```
Comment (simon-frankau) [inline assignment3/WordSearch.java:64]: This works fine. You might consider passing an accumulator collection variable around to add items to, rather than returning a new collection for each recursive call, and copying over the sets each time.

**Review by simon-frankau — APPROVED** [review]
Algorithm could do with a couple of minor tweaks, but is basically fine. Slight code style improvements and improved testing would be appreciated.

What's the algorithmic complexity of your algorithm?
